### ISSUE TITLE
WhenAll exits early when any task has an exception

### ISSUE THREAD

**Issue by all-iver** [Cysharp/UniTask#514]
It seems as though WhenAll immediately fails with an exception if any awaited task throws an exception, instead of waiting for all the tasks to complete and then throwing the exception.  I believe this is different than normal Task behavior which would wait until all tasks are completed regardless of whether some of them failed.

This was unexpected behavior for me because I am expecting some tasks to fail occasionally, but I still need to know that all tasks are done before continuing.  I understand from #335 that WhenAll doesn't store the exceptions for individual tasks, but I think it should not convert to WhenAny just because one task failed...perhaps it could store the first exception that happened and then re-throw that when all tasks are done?

Or, is there some way I'm missing to wait for all tasks to complete regardless of failure?

**Comment by neuecc** [comment]
Yes, the reason WhenAll behaves differently than Task.WhenAll is due to a failure in my initial design.
I would like to fix it, but am having trouble thinking of a way to do so because it would break compatibility.

**Comment by timcassell** [comment]
If it helps, I recently solved this problem in my promise library by adding `Promise.AllSettled` to match the es6 method (https://github.com/timcassell/ProtoPromise#allsettled-and-mergesettled). You could do the same by adding `WhenAllSettled`.

**Comment by andreiagmu** [comment]
Has anyone checked out this solution? (by @adammyhre)
https://gist.github.com/adammyhre/40e7c968edda098fdebb2adbe806c911

Source, see `Important Notes 4)` in the video description:
https://www.youtube.com/watch?v=OlNJ2GkaPTE

**Comment by hmkc** [comment]
It might be possible to convert UniTask to Task using AsTask(), then use Task.WhenAll() to replace UniTask.WhenAll()

**Comment by neuecc** [comment]
I will never change behavior using `#define`.
If you want to add it, you should prepare it as a separate method name and then use an Analyzer to prohibit the conventional WhenAll.
Implementing it yoruself should only require small adjustments if you reference the existing implementation.

**Comment by johntday** [comment]
What is the recommended approach from Cysharp ?

**Comment by imniko** [comment]
Amazing to see this issue is not handled for more almost 3 year now.

**Comment by RichardWepner** [comment]
> Amazing to see this issue is not handled for more almost 3 year now.

@imniko: @neuecc was responding already that they have difficulties coming up with a good solution to this. If you're interested in this being fixed, it might be helpful if you could take a code that's currently present and present a solution to this, possibly through a Pull Request (i.e. with the solution being implemented already in a verifiable/testable form).
